dor/cli/upload.py

```diff
@@ -39,38 +43,54 @@ async def _run_upload(
-            fileset_profiles[name] = {}    
-        fileset_profiles[name][file_name] = profiles[file_name]    
-    for name in fileset_profiles:             
-        try:
-            async with httpx.AsyncClient(follow_redirects=True) as client:
-                result = await run_upload_fileset(
-                    client,
-                    base_url,
-                    profiles=fileset_profiles[name],
-                    folder=folder,
-                    name=name,
-                    project_id=project_id,                   
-                )
-            typer.echo(f"Fileset created successfully: {result}")
+            fileset_profiles[name] = {}
+        fileset_profiles[name][file_name] = profiles[file_name]
 
-        except UploadError as exc:
-            typer.secho(f"Upload error: {exc}", fg="white", bg="red", bold=True)
-            typer.echo(f"Details: {exc.message}, Code: {exc.code}", err=True)
-            raise typer.Exit(1)
+    semaphore = asyncio.Semaphore(10)
 
-        except httpx.RequestError as exc:
-            typer.echo(f"An error occurred while making the request: {exc}", err=True)
-            raise typer.Exit(1)
+    async with httpx.AsyncClient(follow_redirects=True) as client:
+        tasks = []
+        for name in fileset_profiles:
+            tasks.append(run_upload_fileset_with_limit(
+                sempahore=semaphore,
+                client=client,
+                base_url=base_url,
+                project_id=project_id,
+                name=name,
+                profiles=fileset_profiles[name],
+                folder=folder,
+            ))
+        results = await asyncio.gather(*tasks, return_exceptions=True)
 
-        except httpx.HTTPStatusError as exc:
+    exceptions = []
+    data_results = []
+    for result in results:
+        if isinstance(result, Exception):
+            exceptions.append(exceptions)
+        else:
+            data_results.append(data_results)
+
+    typer.echo(f"{len(data_results)} filesets created successfully: {data_results}")
+
+    for exception in exceptions:
+        if isinstance(exception, UploadError):
+            typer.secho(f"Upload error: {exception}", fg="white", bg="red", bold=True)
+            typer.echo(f"Details: {exception.message}, Code: {exception.code}", err=True)
+        elif isinstance(exception, httpx.RequestError):
+            typer.echo(f"An error occurred while making a request: {exception}", err=True)
+        elif isinstance(exception, httpx.HTTPStatusError):
             typer.echo( 
-                f"HTTP error occurred: {exc.response.status_code} - {exc.response.text}",
+                f"HTTP error occurred: {exception.response.status_code} - {exception.response.text}",
                 err=True,
             )
-            raise typer.Exit(1)
+        else:
+            typer.echo("Unknown exception occurred: {exception}")
+
+    if exceptions:
+        raise typer.Exit(1)
```

Comment:
How do we want to handle possible exceptions? The current approach assumes that all requests should be executed, and exceptions are gathered and then reported. We could alternatively use a [`TaskGroup`](https://docs.python.org/3/library/asyncio-task.html#asyncio.TaskGroup), which cancels any remaining work once an error is encountered. I lean towards the former behavior, but I could see some situations where the latter is preferred.

Reply:
Could you add a question about this to https://github.com/mlibrary/project-dor-docs/discussions/89? This intersects with whether upload detaches the process, and whether the DCS user configures the upload to continue or exit.